plugins/single_plugins/preserve-output.cpp

```diff
@@ -0,0 +1,314 @@
+         "") || core_focused_output_expired())
+    {
+        LOGD("Setting last focused output to: ", output->to_string());
+        wf::get_core().get_data_safe<last_focused_output_t>()->output_identifier =
+            make_output_identifier(output);
+        wf::get_core().get_data_safe<last_focused_output_t>()->timestamp =
+            std::chrono::steady_clock::now();
+    }
+}
+
+std::string core_get_focused_output()
+{
+    return wf::get_core().get_data_safe<last_focused_output_t>()->output_identifier;
+}
+
+void core_erase_focused_output()
+{
+    wf::get_core().erase_data<last_focused_output_t>();
+}
+
+/**
+ * preserve-output plugin
+ */
+
+class wayfire_preserve_output : public wf::plugin_interface_t
+{
+    std::vector<std::string> outputs_being_removed;
+
+    wf::signal_connection_t output_pre_remove = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_pre_remove_signal*)data;
+        LOGD("Received pre-remove event: ", signal_data->output->to_string());
+        outputs_being_removed.push_back(signal_data->output->to_string());
+        LOGD("Outputs currently being removed: ");
+        for (auto el : outputs_being_removed)
+        {
+            LOGD("- ", el);
+        }
+
+        if (signal_data->output != output)
+        {
+            // This event is not for this output
+            return;
+        }
+
+        // This output is being destroyed
+        std::string identifier = make_output_identifier(output);
+
+        // Store this output as the focused one
+        if (wf::get_core().get_active_output()->to_string() == output->to_string())
+        {
+            core_store_focused_output(output);
+        }
+
+        auto views = output->workspace->get_views_in_layer(wf::LAYER_WORKSPACE);
+        for (uint i = 0; i < views.size(); i++)
+        {
+            auto view = views[i];
+            if ((view->role != wf::VIEW_ROLE_TOPLEVEL) || !view->is_mapped())
+            {
+                continue;
+            }
+
+            // Set current output and geometry in the view's last output data
+            if (!view_has_data(view))
+            {
+                view_store_data(view, output, i);
+            }
+        }
+    };
+
+    wf::signal_connection_t output_removed = [=] (wf::signal_data_t *data)
+    {
+        auto signal_data = (wf::output_removed_signal*)data;
+        LOGD("Received output-removed event: ", signal_data->output->to_string());
+        outputs_being_removed.erase(std::remove(
+            outputs_being_removed.begin(),
+            outputs_being_removed.end(),
+            signal_data->output->to_string()), outputs_being_removed.end());
+        LOGD("Outputs currently being removed: ");
+        for (auto el : outputs_being_removed)
+        {
+            LOGD("- ", el);
+        }
+    };
+
+    wf::signal_connection_t restore_views_to_output = [=] (wf::signal_data_t *data)
```

Comment:
I thought the plan was to do that only if the data has not expired yet? And we need a configuration option for the expiration time for sure. It should be easy enough to add, just copy any existing .xml file in `metadata`, and create a new option with type int and default value `10000`.

Reply:
The data expiration is only for the last output focusing. The pseudocode is:

```
core:
    last_focused_output (identifier)
    timestamp

output pre-destroy:
    if output is focused:
        if no core.last_focused_output set or core.timestamp is more than X ms ago:
            core.last_focused_output = this output identifier 
            core.timestamp = now 
                                   
output added:                      
    if core is this output: 
        focus this output          
        delete core.last_focused_output  
```

As you see, the data expiration is for when an output sets itself as the last focused - it won't set itself as the last focused output if another output has already done so in the last 10 seconds. This is because when we destroy the output that has focus, the other output automatically gets the focus, so when the second output is also destroyed, it will also think it's the one with focus. This way we remember which output actually had focus before the output destruction started. The 10 second timeout is so if someone actually disconnects an output, it isn't stored as the last focused output forever.

Reply:
Ok. So we'll always restore the views on their orignal output (if they aren't moved) even if the output is plugged after 10 days? (I have nothing against, but I want to know what the desired behavior is)

Reply:
Yes we will restore the views on their original output no matter how long has passed if they haven't been explicitly moved by the user.

Reply:
hmm, ok. thanks for the explanation

Reply:
however it still should be an option ...

Reply:
I've made this a config option: last_output_focus_timeout
Clunky name but it's pretty hard to describe what it does in a few words :)